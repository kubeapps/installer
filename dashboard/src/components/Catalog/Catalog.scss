.empty-catalog {
  text-align: center;
  width: 100%;
  height: calc(100vh - 6rem);

  cds-icon {
    margin-top: 2rem;
    width: 4rem;
    height: 4rem;
  }

  cds-button {
    margin-top: 2rem;
  }

  p {
    margin-top: 0.6rem;
  }
}

.scrollHandler {
  margin-top: 1rem;
  margin-bottom: 1rem;
  min-height: 50vh;
}

.catalogContainer {
  padding-bottom: 5vh;
}

.endPageMessage {
  width: 100%;
  margin-top: 2rem;
  text-align: center;
}

.kubeapps-main-container .content-area {
  overflow-y: scroll;
}
