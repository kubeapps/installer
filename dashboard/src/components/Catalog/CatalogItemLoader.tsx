import React from "react";
import ContentLoader from "react-content-loader";

export interface ICatalogItemLoader {
  loaded?: boolean;
  children?: any;
  innerProps?: string;
}

function CatalogItemLoader(props: ICatalogItemLoader) {
  const { loaded, children, innerProps } = props;
  return loaded ? (
    children
  ) : (
    <ContentLoader
      speed={2}
      width={300}
      height={300}
      className="contentLoader"
      viewBox="0 0 200 200"
      backgroundColor="#dedede"
      foregroundColor="#ffffff"
      {...innerProps}
    >
      <rect x="21" y="30" rx="0" ry="0" width="200" height="10" />
      <circle cx="61" cy="78" r="20" />
      <rect x="110" y="60" rx="0" ry="0" width="98" height="5" />
      <rect x="21" y="111" rx="0" ry="0" width="200" height="19" />
      <rect x="110" y="70" rx="0" ry="0" width="98" height="5" />
      <rect x="110" y="80" rx="0" ry="0" width="98" height="5" />
      <rect x="110" y="90" rx="0" ry="0" width="98" height="5" />

      <rect x="21" y="30" rx="0" ry="0" width="200" height="10" />
      <circle cx="61" cy="78" r="20" />
      <rect x="110" y="60" rx="0" ry="0" width="98" height="5" />
      <rect x="21" y="111" rx="0" ry="0" width="200" height="19" />
      <rect x="110" y="70" rx="0" ry="0" width="98" height="5" />
      <rect x="110" y="80" rx="0" ry="0" width="98" height="5" />
      <rect x="110" y="90" rx="0" ry="0" width="98" height="5" />

      <rect x="21" y="30" rx="0" ry="0" width="200" height="10" />
      <circle cx="61" cy="78" r="20" />
      <rect x="110" y="60" rx="0" ry="0" width="98" height="5" />
      <rect x="21" y="111" rx="0" ry="0" width="200" height="19" />
      <rect x="110" y="70" rx="0" ry="0" width="98" height="5" />
      <rect x="110" y="80" rx="0" ry="0" width="98" height="5" />
      <rect x="110" y="90" rx="0" ry="0" width="98" height="5" />
    </ContentLoader>
  );
}

export default CatalogItemLoader;
