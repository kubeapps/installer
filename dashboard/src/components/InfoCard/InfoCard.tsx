import * as React from "react";

import { Link } from "react-router-dom";
import Card, { CardContent, CardIcon } from "../Card";
import "./InfoCard.css";

export interface IServiceInstanceCardProps {
  title: string;
  info: string;
  link?: string;
  icon?: string;
  tag1Class?: string;
  tag1Content?: string;
  tag2Class?: string;
  tag2Content?: string;
}

const InfoCard: React.SFC<IServiceInstanceCardProps> = props => {
  const { title, link, icon, info, tag1Content, tag1Class, tag2Content, tag2Class } = props;
  return (
    <Card responsive={true} className="ListItem">
      <Link to={link || "#"} title={title}>
        <CardIcon icon={icon} />
        <CardContent>
          <div className="ListItem__content">
            <h3 className="ListItem__content__title type-big">{title}</h3>
            <div className="ListItem__content__info">
              <p className="margin-reset type-small padding-t-tiny type-color-light-blue">{info}</p>
              <div>
                {tag1Content && (
                  <span
                    className={`ListItem__content__info_tag ListItem__content__info_tag-1 type-small type-color-white padding-t-tiny padding-h-normal ${tag1Class ||
                      ""}`}
                  >
                    {tag1Content}
                  </span>
                )}
                {tag2Content && (
                  <span
                    className={`ListItem__content__info_tag ListItem__content__info_tag-2 type-small type-color-white padding-t-tiny padding-h-normal margin-b-normal ${tag2Class ||
                      ""}`}
                  >
                    {tag2Content}
                  </span>
                )}
              </div>
            </div>
          </div>
        </CardContent>
      </Link>
    </Card>
  );
};

export default InfoCard;
