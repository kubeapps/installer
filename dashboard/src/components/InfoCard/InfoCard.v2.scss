.card-header {
  img {
    max-width: 24px;
    float: right;
  }
}

.card-block {
  height: 4.5rem;
  overflow-y: hidden;
  .card-icon {
    text-align: center;
    img {
      max-height: 3.5rem;
      max-width: 100%;
    }
  }
}

.card-footer {
  height: 3.6rem;
}

.card.clickable a {
  text-decoration: none;
  color: black;
}

.footer-tags {
  text-align: right;
}
