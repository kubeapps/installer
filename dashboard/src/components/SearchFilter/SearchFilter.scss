#search {
  color: var(--cds-global-typography-color-500, #000);
  margin-left: 0.3rem;
}

#search input::placeholder {
  color: var(--cds-global-typography-color-300, #454545);
}

header .search-box form {
  border-radius: 0.15rem;
  margin-left: 1rem;
}
