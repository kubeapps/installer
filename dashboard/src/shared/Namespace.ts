import { axios } from "./Auth";

import { IK8sList, IResource } from "./types";

export default class Namespace {
  public static async list() {
    const { data } = await axios.get<IK8sList<IResource, {}>>(`${Namespace.APIEndpoint}`);
    return data;
  }

  private static APIBase: string = "/api/kube";
  private static APIEndpoint: string = `${Namespace.APIBase}/api/v1/namespaces`;
}
