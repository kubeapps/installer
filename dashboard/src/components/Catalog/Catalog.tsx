import { CdsButton } from "@clr/react/button";
import { CdsIcon } from "@clr/react/icon";
import FilterGroup from "components/FilterGroup/FilterGroup";
import Alert from "components/js/Alert";
import Column from "components/js/Column";
import Row from "components/js/Row";
import { push } from "connected-react-router";
import { flatten, get, intersection, uniq, without } from "lodash";
import React, { useEffect, useState } from "react";
import { useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import { definedNamespaces } from "shared/Namespace";
import { app } from "shared/url";
import { IChartState, IClusterServiceVersion } from "../../shared/types";
import { escapeRegExp } from "../../shared/utils";
import LoadingWrapper from "../LoadingWrapper/LoadingWrapper";
import PageHeader from "../PageHeader/PageHeader";
import SearchFilter from "../SearchFilter/SearchFilter";
import "./Catalog.css";
import CatalogItems from "./CatalogItems";

function categoryToReadable(category: string) {
  return category === "" ? "Unknown" : category.replace(/([a-z])([A-Z][a-z])/g, "$1 $2").trimLeft();
}

function getOperatorCategories(c: IClusterServiceVersion): string[] {
  return get(c, "metadata.annotations.categories", "")
    .split(",")
    .map((category: string) => categoryToReadable(category));
}

interface ICatalogProps {
  charts: IChartState;
  repo: string;
  filter: { [name: string]: string };
  fetchCharts: (cluster: string, namespace: string, repo: string) => void;
  cluster: string;
  namespace: string;
  kubeappsNamespace: string;
  getCSVs: (cluster: string, namespace: string) => void;
  csvs: IClusterServiceVersion[];
}

export const filterNames = {
  SEARCH: "Search",
  TYPE: "Type",
  REPO: "Repository",
  CATEGORY: "Category",
  OPERATOR_PROVIDER: "Provider",
};

export function initialFilterState() {
  const result = {};
  Object.values(filterNames).forEach(f => (result[f] = []));
  return result;
}

export function filtersToQuery(filters: any) {
  let query = "";
  const activeFilters = Object.keys(filters).filter(f => filters[f].length);
  if (activeFilters.length) {
    const filterQueries = activeFilters.map(
      filter => `${filter}=${filters[filter].map((f: string) => encodeURIComponent(f)).join(",")}`,
    );
    query = "?" + filterQueries.join("&");
  }
  return query;
}

function Catalog(props: ICatalogProps) {
  const {
    charts: {
      isFetching,
      selected: { error },
      items: charts,
    },
    fetchCharts,
    cluster,
    namespace,
    getCSVs,
    csvs,
    repo,
    filter: propsFilter,
  } = props;
  const dispatch = useDispatch();
  const [filters, setFilters] = useState(initialFilterState());

  useEffect(() => {
    const newFilters = {};
    Object.keys(propsFilter).forEach(filter => {
      newFilters[filter] = propsFilter[filter].split(",");
    });
    setFilters({
      ...initialFilterState(),
      ...newFilters,
    });
  }, [propsFilter]);

  const pushFilters = (newFilters: any) => {
    dispatch(push(app.catalog(cluster, namespace) + filtersToQuery(newFilters)));
  };
  const addFilter = (type: string, value: string) => {
    pushFilters({
      ...filters,
      [type]: filters[type].concat(value),
    });
  };
  const removeFilter = (type: string, value: string) => {
    pushFilters({
      ...filters,
      [type]: without(filters[type], value),
    });
  };
  const removeFilterFunc = (type: string, value: string) => {
    return () => removeFilter(type, value);
  };
  const clearAllFilters = () => {
    pushFilters({});
  };
  const submitFilters = () => {
    pushFilters(filters);
  };

  const allRepos = uniq(charts.map(c => c.attributes.repo.name));
  const allProviders = uniq(csvs.map(c => c.spec.provider.name));
  const allCategories = uniq(
    charts
      .map(c => categoryToReadable(c.attributes.category))
      .concat(flatten(csvs.map(c => getOperatorCategories(c)))),
  ).sort();

  useEffect(() => {
    fetchCharts(cluster, namespace, repo);
    getCSVs(cluster, namespace);
  }, [cluster, namespace, repo, fetchCharts, getCSVs]);

  // Only one search filter can be set
  const searchFilter = filters[filterNames.SEARCH][0] || "";
  const setSearchFilter = (searchTerm: string) => {
    setFilters({
      ...filters,
      [filterNames.SEARCH]: [searchTerm],
    });
  };

  const filteredCharts = charts
    .filter(
      () => filters[filterNames.TYPE].length === 0 || filters[filterNames.TYPE].includes("Charts"),
    )
    .filter(() => filters[filterNames.OPERATOR_PROVIDER].length === 0)
    .filter(c => new RegExp(escapeRegExp(searchFilter), "i").test(c.id))
    .filter(
      c =>
        filters[filterNames.REPO].length === 0 ||
        filters[filterNames.REPO].includes(c.attributes.repo.name),
    )
    .filter(
      c =>
        filters[filterNames.CATEGORY].length === 0 ||
        filters[filterNames.CATEGORY].includes(categoryToReadable(c.attributes.category)),
    );
  const filteredCSVs = csvs
    .filter(
      () =>
        filters[filterNames.TYPE].length === 0 || filters[filterNames.TYPE].includes("Operators"),
    )
    .filter(() => filters[filterNames.REPO].length === 0)
    .filter(c => new RegExp(escapeRegExp(searchFilter), "i").test(c.metadata.name))
    .filter(
      c =>
        filters[filterNames.OPERATOR_PROVIDER].length === 0 ||
        filters[filterNames.OPERATOR_PROVIDER].includes(c.spec.provider.name),
    )
    .filter(
      c =>
        filters[filterNames.CATEGORY].length === 0 ||
        intersection(filters[filterNames.CATEGORY], getOperatorCategories(c)).length,
    );

  return (
    <section>
      <PageHeader
        title="Catalog"
        filter={
          <SearchFilter
            key="searchFilter"
            placeholder="search charts..."
            onChange={setSearchFilter}
            value={searchFilter}
            submitFilters={submitFilters}
          />
        }
      />
      <LoadingWrapper loaded={!isFetching}>
        {error && (
          <Alert theme="danger">
            An error occurred while fetching the catalog: {error.message}
          </Alert>
        )}
        {namespace === definedNamespaces.all ? (
          <div className="empty-catalog">
            <CdsIcon shape="file-group" />
            <h4>A valid namespace should be selected.</h4>
            <p>
              In order to show the catalog of applications available, a namespace must be selected
              using the selector in the top right corner.
            </p>
          </div>
        ) : charts.length === 0 && csvs.length === 0 ? (
          <div className="empty-catalog">
            <CdsIcon shape="bundle" />
            <p>The current catalog is empty.</p>
            <p>
              Manage your Helm chart repositories in Kubeapps by visiting the App repositories
              configuration page.
            </p>
            <Link to={app.config.apprepositories(cluster, namespace)}>
              <CdsButton>Manage App Repositories</CdsButton>
            </Link>
          </div>
        ) : (
          <Row>
            <Column span={2}>
              <div className="filters-menu">
                <h5>
                  Filters{" "}
                  {flatten(Object.values(filters)).length ? (
                    <CdsButton size="sm" action="flat" onClick={clearAllFilters}>
                      Clear All
                    </CdsButton>
                  ) : (
                    <></>
                  )}{" "}
                </h5>
                {csvs.length > 0 && (
                  <div className="filter-section">
                    <label>Application Type</label>
                    <FilterGroup
                      name={filterNames.TYPE}
                      options={["Operators", "Charts"]}
                      currentFilters={filters[filterNames.TYPE]}
                      onAddFilter={addFilter}
                      onRemoveFilter={removeFilter}
                    />
                  </div>
                )}
                {allCategories.length > 0 && (
                  <div className="filter-section">
                    <label className="filter-label">Category</label>
                    <FilterGroup
                      name={filterNames.CATEGORY}
                      options={allCategories}
                      currentFilters={filters[filterNames.CATEGORY]}
                      onAddFilter={addFilter}
                      onRemoveFilter={removeFilter}
                    />
                  </div>
                )}
                {allRepos.length > 0 && (
                  <div className="filter-section">
                    <label>Application Repository</label>
                    <FilterGroup
                      name={filterNames.REPO}
                      options={allRepos}
                      currentFilters={filters[filterNames.REPO]}
                      onAddFilter={addFilter}
                      onRemoveFilter={removeFilter}
                    />
                  </div>
                )}
                {allProviders.length > 0 && (
                  <div className="filter-section">
                    <label className="filter-label">Operator Provider</label>
                    <FilterGroup
                      name={filterNames.OPERATOR_PROVIDER}
                      options={allProviders}
                      currentFilters={filters[filterNames.OPERATOR_PROVIDER]}
                      onAddFilter={addFilter}
                      onRemoveFilter={removeFilter}
                    />
                  </div>
                )}
              </div>
            </Column>
            <Column span={10}>
              <>
                <div className="filter-summary">
                  {Object.keys(filters).map(filterName => {
                    if (filters[filterName].length) {
                      return filters[filterName].map((filterValue: string, i: number) => (
                        <span key={`${filterName}-${filterValue}`} className="label label-info">
                          {filterName}: {filterValue}{" "}
                          <CdsIcon
                            shape="times"
                            onClick={removeFilterFunc(filterName, filterValue)}
                          />
                        </span>
                      ));
                    }
                    return null;
                  })}
                </div>
                <Row>
                  <CatalogItems
                    charts={filteredCharts}
                    csvs={filteredCSVs}
                    cluster={cluster}
                    namespace={namespace}
                  />
                </Row>
              </>
            </Column>
          </Row>
        )}
      </LoadingWrapper>
    </section>
  );
}

export default Catalog;
