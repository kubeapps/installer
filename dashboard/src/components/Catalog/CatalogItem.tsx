import * as React from "react";
import { Link } from "react-router-dom";

import placeholder from "../../placeholder.png";
import { IChart } from "../../shared/types";
import InfoCard from "../InfoCard";

import "./CatalogItem.css";

interface ICatalogItemProps {
  chart: IChart;
}

class CatalogItem extends React.Component<ICatalogItemProps> {
  public render() {
    const { chart } = this.props;
    const { icon, name, repo } = chart.attributes;
    const iconSrc = icon ? `/api/chartsvc/${icon}` : placeholder;
    const latestAppVersion = chart.relationships.latestChartVersion.data.app_version;
    const repoTag = (
      <Link className="ListItem__content__info_tag_link" to={`/catalog/${repo.name}`}>
        {repo.name}
      </Link>
    );
    return (
      <InfoCard
        key={`${repo}/${name}`}
        title={name}
        link={`/charts/${chart.id}`}
        info={latestAppVersion || "-"}
        icon={iconSrc}
        tag1Content={repoTag}
        tag1Class={repo.name}
      />
    );
  }
}

export default CatalogItem;
