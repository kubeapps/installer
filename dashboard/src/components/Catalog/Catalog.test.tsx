import * as React from "react";

import FilterGroup from "components/FilterGroup/FilterGroup";
import InfoCard from "components/InfoCard/InfoCard";
import Alert from "components/js/Alert";
import { act } from "react-dom/test-utils";
import { definedNamespaces } from "shared/Namespace";
import { defaultStore, getStore, mountWrapper } from "shared/specs/mountWrapper";
import { IChart, IChartState, IClusterServiceVersion } from "../../shared/types";
import SearchFilter from "../SearchFilter/SearchFilter";
import Catalog, { filterNames } from "./Catalog";

const defaultChartState = {
  isFetching: false,
  selected: {} as IChartState["selected"],
  deployed: {} as IChartState["deployed"],
  items: [],
  updatesInfo: {},
} as IChartState;
const defaultProps = {
  charts: defaultChartState,
  repo: "",
  filter: {},
  fetchCharts: jest.fn(),
  pushSearchFilter: jest.fn(),
  cluster: "default",
  namespace: "kubeapps",
  kubeappsNamespace: "kubeapps",
  csvs: [],
  getCSVs: jest.fn(),
};
const chartItem = {
  id: "foo",
  attributes: {
    name: "foo",
    description: "",
    category: "",
    repo: { name: "foo", namespace: "chart-namespace" },
  },
  relationships: { latestChartVersion: { data: { app_version: "v1.0.0" } } },
} as IChart;
const chartItem2 = {
  id: "bar",
  attributes: {
    name: "bar",
    description: "",
    category: "Database",
    repo: { name: "bar", namespace: "chart-namespace" },
  },
  relationships: { latestChartVersion: { data: { app_version: "v2.0.0" } } },
} as IChart;
const csv = {
  metadata: {
    name: "test-csv",
  },
  spec: {
    provider: {
      name: "me",
    },
    icon: [{ base64data: "data", mediatype: "img/png" }],
    customresourcedefinitions: {
      owned: [
        {
          name: "foo-cluster",
          displayName: "foo-cluster",
          version: "v1.0.0",
          description: "a meaningful description",
        },
      ],
    },
  },
} as IClusterServiceVersion;
const populatedProps = {
  ...defaultProps,
  csvs: [csv],
  charts: { ...defaultChartState, items: [chartItem, chartItem2] },
};

it("retrieves csvs in the namespace", () => {
  const getCSVs = jest.fn();
  mountWrapper(defaultStore, <Catalog {...populatedProps} getCSVs={getCSVs} />);
  expect(getCSVs).toHaveBeenCalledWith(defaultProps.cluster, defaultProps.namespace);
});

it("shows all the elements", () => {
  const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} />);
  expect(wrapper.find(InfoCard)).toHaveLength(3);
});

it("should render a message if there are no elements in the catalog", () => {
  const wrapper = mountWrapper(defaultStore, <Catalog {...defaultProps} />);
  const message = wrapper.find(".empty-catalog");
  expect(message).toExist();
  expect(message).toIncludeText("The current catalog is empty");
});

it("should render an error if it exists", () => {
  const charts = {
    ...defaultChartState,
    selected: {
      error: new Error("Boom!"),
    },
  } as any;
  const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} charts={charts} />);
  const error = wrapper.find(Alert);
  expect(error.prop("theme")).toBe("danger");
  expect(error).toIncludeText("Boom!");
});

it("behaves like a loading wrapper", () => {
  const wrapper = mountWrapper(
    defaultStore,
    <Catalog {...populatedProps} charts={{ isFetching: true, items: [], selected: {} } as any} />,
  );
  expect(wrapper.find("LoadingWrapper")).toExist();
});

it("shows a message if the namespace has not been selected", () => {
  const wrapper = mountWrapper(
    defaultStore,
    <Catalog {...populatedProps} namespace={definedNamespaces.all} />,
  );
  expect(wrapper).toIncludeText("A valid namespace should be selected");
  expect(wrapper.find(".filters-menu")).not.toExist();
});

describe("filters by the searched item", () => {
  it("filters using prop", () => {
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog {...populatedProps} filter={{ Search: "bar" }} />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });

  it("filters modifying the search box", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} />);
    act(() => {
      (wrapper.find(SearchFilter).prop("onChange") as any)("bar");
    });
    wrapper.update();
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });
});

describe("filters by application type", () => {
  it("doesn't show the filter if there are no csvs", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...populatedProps} csvs={[]} />);
    expect(
      wrapper.find(FilterGroup).findWhere(g => g.prop("name") === filterNames.TYPE),
    ).not.toExist();
  });

  it("filters only charts", () => {
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog {...populatedProps} filter={{ Type: "Charts" }} />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(2);
  });

  it("push filter for only charts", () => {
    const store = getStore({});
    const wrapper = mountWrapper(store, <Catalog {...populatedProps} />);
    const input = wrapper.find("input").findWhere(i => i.prop("value") === "Charts");
    input.simulate("change", { target: { value: "Charts" } });
    // It should have pushed with the filter
    expect(store.getActions()[0].payload).toEqual({
      args: ["/c/default/ns/kubeapps/catalog?Type=Charts"],
      method: "push",
    });
  });

  it("filters only operators", () => {
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog {...populatedProps} filter={{ Type: "Operators" }} />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });

  it("push filter for only operators", () => {
    const store = getStore({});
    const wrapper = mountWrapper(store, <Catalog {...populatedProps} />);
    const input = wrapper.find("input").findWhere(i => i.prop("value") === "Operators");
    input.simulate("change", { target: { value: "Operators" } });
    // It should have pushed with the filter
    expect(store.getActions()[0].payload).toEqual({
      args: ["/c/default/ns/kubeapps/catalog?Type=Operators"],
      method: "push",
    });
  });
});

describe("filters by application repository", () => {
  it("doesn't show the filter if there are no apps", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...defaultProps} />);
    expect(
      wrapper.find(FilterGroup).findWhere(g => g.prop("name") === filterNames.REPO),
    ).not.toExist();
  });

  it("filters by repo", () => {
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog {...populatedProps} filter={{ [filterNames.REPO]: "foo" }} />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });

  it("push filter for repo", () => {
    const store = getStore({});
    const wrapper = mountWrapper(store, <Catalog {...populatedProps} />);
    // The repo name is "foo"
    const input = wrapper.find("input").findWhere(i => i.prop("value") === "foo");
    input.simulate("change", { target: { value: "foo" } });
    // It should have pushed with the filter
    expect(store.getActions()[0].payload).toEqual({
      args: ["/c/default/ns/kubeapps/catalog?Repository=foo"],
      method: "push",
    });
  });
});

describe("filters by operator provider", () => {
  it("doesn't show the filter if there are no csvs", () => {
    const wrapper = mountWrapper(defaultStore, <Catalog {...defaultProps} />);
    expect(
      wrapper.find(FilterGroup).findWhere(g => g.prop("name") === filterNames.OPERATOR_PROVIDER),
    ).not.toExist();
  });

  const csv2 = {
    metadata: {
      name: "csv2",
    },
    spec: {
      ...csv.spec,
      provider: {
        name: "you",
      },
    },
  } as any;

  it("push filter for operator provider", () => {
    const store = getStore({});
    const wrapper = mountWrapper(store, <Catalog {...populatedProps} csvs={[csv, csv2]} />);
    const input = wrapper.find("input").findWhere(i => i.prop("value") === "you");
    input.simulate("change", { target: { value: "you" } });
    // It should have pushed with the filter
    expect(store.getActions()[0].payload).toEqual({
      args: ["/c/default/ns/kubeapps/catalog?Provider=you"],
      method: "push",
    });
  });

  it("filters by operator provider", () => {
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog
        {...populatedProps}
        csvs={[csv, csv2]}
        filter={{ [filterNames.OPERATOR_PROVIDER]: "you" }}
      />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });
});

describe("filters by category", () => {
  it("renders a Unknown category if not set", () => {
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog {...defaultProps} charts={{ ...defaultChartState, items: [chartItem] }} />,
    );
    expect(wrapper.find("input").findWhere(i => i.prop("value") === "Unknown")).toExist();
  });

  it("push filter for category", () => {
    const store = getStore({});
    const wrapper = mountWrapper(
      store,
      <Catalog
        {...defaultProps}
        charts={{ ...defaultChartState, items: [chartItem, chartItem2] }}
      />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(2);
    const input = wrapper.find("input").findWhere(i => i.prop("value") === "Database");
    input.simulate("change", { target: { value: "Database" } });
    // It should have pushed with the filter
    expect(store.getActions()[0].payload).toEqual({
      args: ["/c/default/ns/kubeapps/catalog?Category=Database"],
      method: "push",
    });
  });

  it("filters a category", () => {
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog
        {...defaultProps}
        charts={{ ...defaultChartState, items: [chartItem, chartItem2] }}
        filter={{ [filterNames.CATEGORY]: "Database" }}
      />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });

  it("filters an operator category", () => {
    const csvWithCat = {
      ...csv,
      metadata: {
        name: "csv-cat",
        annotations: {
          categories: "E-Learning",
        },
      },
    } as any;
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog
        {...defaultProps}
        csvs={[csv, csvWithCat]}
        filter={{ [filterNames.CATEGORY]: "E-Learning" }}
      />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });

  it("filters operator categories", () => {
    const csvWithCat = {
      ...csv,
      metadata: {
        name: "csv-cat",
        annotations: {
          categories: "DeveloperTools, Infrastructure",
        },
      },
    } as any;
    const wrapper = mountWrapper(
      defaultStore,
      <Catalog
        {...defaultProps}
        csvs={[csv, csvWithCat]}
        filter={{ [filterNames.CATEGORY]: "Developer Tools,Infrastructure" }}
      />,
    );
    expect(wrapper.find(InfoCard)).toHaveLength(1);
  });
});
