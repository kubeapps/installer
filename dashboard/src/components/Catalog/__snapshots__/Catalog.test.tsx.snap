// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderization when charts available should render the list of charts 1`] = `
<section
  className="Catalog"
>
  <PageHeader>
    <h1>
      Catalog
    </h1>
    <SearchFilter
      className="margin-l-big"
      onChange={[Function]}
      onSubmit={[Function]}
      placeholder="search charts..."
      value=""
    />
  </PageHeader>
  <LoadingWrapper
    loaded={true}
    type={0}
  >
    <div
      className="row"
    >
      <div
        className=""
      >
        <CardGrid>
          <CatalogItem
            item={
              Object {
                "description": "",
                "icon": undefined,
                "id": "bar",
                "name": "bar",
                "namespace": "kubeapps",
                "repoName": "bar",
                "type": "chart",
                "version": "v2.0.0",
              }
            }
            key="chart/bar/bar"
          />
          <CatalogItem
            item={
              Object {
                "description": "",
                "icon": undefined,
                "id": "foo",
                "name": "foo",
                "namespace": "kubeapps",
                "repoName": "foo",
                "type": "chart",
                "version": "v1.0.0",
              }
            }
            key="chart/foo/foo"
          />
        </CardGrid>
      </div>
    </div>
  </LoadingWrapper>
</section>
`;

exports[`renderization when fetching apps loading spinner matches the snapshot 1`] = `
<section
  className="Catalog"
>
  <PageHeader>
    <h1>
      Catalog
    </h1>
    <SearchFilter
      className="margin-l-big"
      onChange={[Function]}
      onSubmit={[Function]}
      placeholder="search charts..."
      value=""
    />
  </PageHeader>
  <LoadingWrapper
    loaded={false}
    type={0}
  >
    <div
      className="row"
    >
      <div
        className=""
      >
        <CardGrid />
      </div>
    </div>
  </LoadingWrapper>
</section>
`;

exports[`renderization when no charts should render an error 1`] = `
<MessageAlertPage
  level="warning"
>
  <div>
    <h5>
      Charts not found.
    </h5>
    Manage your Helm chart repositories in Kubeapps by visiting the
     
    <Link
      to="/config/repos"
    >
      App repositories configuration
    </Link>
     page.
  </div>
</MessageAlertPage>
`;
